evalml/automl/progress.py

```diff
@@ -0,0 +1,172 @@
+        else:
+            self.logger = logging.getLogger(__name__)
+
+    def should_continue(self, results, interrupted=False, mid_batch=False):
+        """Given AutoML Results, return whether or not the search should continue.
+
+        Args:
+            results (dict): AutoMLSearch results.
+            interrupted (bool): whether AutoMLSearch was given an keyboard interrupt.
+            mid_batch (bool): whether this method was called while in the middle of a batch or not.
+
+        Returns:
+            bool: True if search should continue, False otherwise.
+        """
+        if interrupted:
+            return False
+        # update and check max_time, max_iterations, and max_batches
+        self.current_time = time.time()
+        self.current_iterations = len(results["pipeline_results"])
+        self.current_batches = self.automl_algorithm.batch_number
+
+        elapsed = self.current_time - self._start_time
+
+        if self.max_time and elapsed >= self.max_time:
+            return False
+        elif self.max_iterations and self.current_iterations >= self.max_iterations:
+            return False
+        elif (
+            self.max_batches
+            and self.current_batches >= self.max_batches
+            and not mid_batch
+        ):
+            return False
+
+        # check for early stopping
+        if self.patience is None or self.tolerance is None:
+            return True
+
+        first_id = results["search_order"][0]
+        best_score = results["pipeline_results"][first_id]["mean_cv_score"]
+        num_without_improvement = 0
+        for id in results["search_order"][1:]:
+            curr_score = results["pipeline_results"][id]["mean_cv_score"]
+            significant_change = (
+                abs((curr_score - best_score) / best_score) > self.tolerance
+            )
+            score_improved = (
+                curr_score > best_score
+                if self.objective.greater_is_better
+                else curr_score < best_score
+            )
+            if score_improved and significant_change:
+                best_score = curr_score
+                num_without_improvement = 0
+            else:
+                num_without_improvement += 1
+            if num_without_improvement >= self.patience:
+                self.logger.info(
+                    "\n\n{} iterations without improvement. Stopping search early...".format(
+                        self.patience,
+                    ),
+                )
+                return False
+        return True
+
+    def _build_progress_dict(self, stopping_criteria, current_state, end_state, unit):
+        progress_dict = {}
+        progress_dict["stopping_criteria"] = stopping_criteria
+        progress_dict["current_state"] = current_state
+        progress_dict["end_state"] = end_state
+        progress_dict["unit"] = unit
+        return progress_dict
+
+    def return_progress(self):
```

Comment:
most of this PR is refactoring to use `should_continue` in this class but this is the important method. Key questions being: is this sufficient information and is the formatting easy to work with to gather information on search progress?

Reply:
Unless I'm missing something more specific, I think we should return progress as a single dictionary rather than as a list of dictionaries. This is mainly due to the large amount of overlap exists between these conditions. We see this with the max_batches and max_iterations logic below. Another example is that I would like to know the elasped time regardless of the stopping criteria.

As an example, I think this function should return the below:
```
{
   "stopping_criteria": "max_batches"
   "elasped_time": self.elasped(),
   "elasped_time_unit": "seconds"
   "max_time": self.max_time, # So None in this case
    "max_batches": self.max_batches, 
   ...
}
```